Add example of `apply_ufunc`

jemmajeffree opened the pull request:
I've written up an example of how to calculate the trend and significance at each gridpoint using xr.apply_ufunc to pass things through to scipy.stats.linregress, so it's vectorised and much faster than using a for loop

Is this something that's of use to the COSIMA community?

ashjbarnes commented:
This looks great! I've added a comment to suggest a bit more motivation at the top of the notebook :)

ashjbarnes commented on Examples/Apply_scipy_function_to_every_gridpoint.ipynb:
```diff
@@ -0,0 +1,533 @@
+{
```
<p>This is great! Maybe a bit more text to help out a beginner? Eg what is vectorising / why do we care.</p><p> </p><p>"Imagine you want to apply this linear regression to every point in space. Looping over every x and y would do the trick but is <em>very slow</em>. Instead, a vectorised function avoids some of the python slowness by employing the fast numpy backend. The <code>xr.apply_ufunc</code> function lets us take a function, and apply it over our whole domain. ... " </p><p> </p><p>I'm tired so that's probably not written very well but you get the gist </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/COSIMA/cosima-recipes/pull/421/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

navidcy replied:
Yes. I like this suggestion. Perhaps have some pseudo code how would you do it with nested for loops and then say that this would be detrimental for performance so we want apply_ufunc?

navidcy commented on Examples/Apply_scipy_function_to_every_gridpoint.ipynb:
```diff
@@ -0,0 +1,533 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #24.    years_since_start = xr.DataArray(np.arange(sst.time.shape[0])/12,
</pre><p> </p><p>this reads hardcoded... what does</p><p> </p>

<pre class="ql-syntax" spellcheck="false">np.arange(sst.time.shape[0])/12
</pre><p> </p><p>mean to do? will it work in general? only with monthly data?</p><p> </p><p>Could we do this more generally?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/COSIMA/cosima-recipes/pull/421/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

navidcy replied:
<pre class="ql-syntax" spellcheck="false">Line #3.    plt.contourf(p_value.xt_ocean,p_value.yt_ocean,p_value,levels=(0,0.05),colors='None',hatches=('...',))
</pre><p>can we have spaces after commas? that's a general comment but it's quite prominent here.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/COSIMA/cosima-recipes/pull/421/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

anton-seaice commented:
Thankyou for going this Jemma!

I think this could go in Tutorials rather than Examples ?

(I will also point out xarray has a [polyfit](https://docs.xarray.dev/en/stable/generated/xarray.Dataset.polyfit.html) which does the same as the notebook :) ) 

If we move to the intake catalog, then the chunking warnings will go away :)

And it would be good to fix up the figure to be tripole aware ... although I guess its not actually essential.

navidcy reviewed (approved):
@jemmajeffree thanks!

I pushed few edits... space after commas etc

the PR looks good to me! I run the notebook and it's all good!

I just don't understand what `np.arange(sst.time.shape[0])/12` does...
Seems something hardcoded to me... something related with months (or the conspicuous coincidental 12)... but I have no idea.

If we can reduce the hardcoding it'd be great. Otherwise a comment next to it explaining what's happening will be useful for somebody reading this example down the road. But I am happy to merge this even as is rather than keep ruminating over details!

navidcy commented:
Also @ashjbarnes [suggestion](https://github.com/COSIMA/cosima-recipes/pull/421#discussion_r1672081175) is nice. If you can incorporate this before merging it'd be nice!

navidcy commented:
@jemmajeffree I think I'll include a sentence along what @ashjbarnes suggested and merge. Unless you plan to edit/work on this?

navidcy commented:
(We can always open a new PR in the future and revisit-enhance the recipe!)

jemmajeffree commented:
I've updated the example here /g/data/x77/jj8842/cosima-recipes/Examples/Apply_scipy_function_to_every_gridpoint.ipynb, but I can't work out how to push it back to the PR and merge with existing changes here. I can try follow instructions, or I'm very happy to have someone else work out how to merge them again.

I've:
      - added some text as suggested by @ashjbarnes
      - tried to clarify what is meant by the np.arange(sst.time.shape[0])/12), which is specific to the linear regression, so I guess hardcoded to this example but I can't see how you'd generalise when you don't know what other functions might be used with this code and what variables they would take

Also noting the comments by @anton-seaice, I know you could do this specific task faster with xr.polyfit, but some things you can't and linregress is an approachable example

Feel free to rewrite any of my explanations, or add anything you think they're lacking

navidcy commented:
We can zoom and I show you?

navidcy commented:
oh I see... probably you can't push because I pushed few edits and you need to merge them! damn... I made it hard....

navidcy commented:
I'll sort it ou!

navidcy commented:
new example -> https://cosima-recipes.readthedocs.io/en/latest/Recipes/Apply_function_to_every_gridpoint.html

thanks @jemmajeffree!

jemmajeffree commented:
Thanks @navidcy for sorting that out!
